### PR TITLE
[BUG, MRG] Quick Bug Fix

### LINKED ISSUE #503: [BUG] Empty dig and line_freq cause errors
```
import os
import os.path as op
import numpy as np

import mne
import mne_bids
from mne.utils import _TempDir

info = mne.create_info(['ch1', 'ch2', 'ch3'], 512,
                       ['seeg'] * 3)
raw_tmp = \
    mne.io.RawArray(np.random.random((3, 1000)) * 1e-6,
                    info)
raw_tmp.info['dig'] = []  # solves issue: raw_tmp.info['dig'] = None
# solves other issue: raw_tmp.info['line_freq'] = 60
out_dir = _TempDir()
raw_tmp.save(op.join(out_dir, 'test.fif'))

raw = mne.io.read_raw_fif(op.join(out_dir, 'test.fif'))

mne_bids.write_raw_bids(raw, 'sub-1_task-mytask', op.join(out_dir, 'bids_dir'),
                            verbose=True, overwrite=True)
```

I got an uninformative index error when my `raw.info['dig']` was an empty list: 
```
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/Users/alexrockhill/software/mne-bids/mne_bids/write.py", line 1175, in write_raw_bids
    raw, kind, overwrite, verbose)
  File "/Users/alexrockhill/software/mne-bids/mne_bids/dig.py", line 304, in _write_dig_bids
    digpoint = raw.info['dig'][0]
IndexError: list index out of range
```

and `PowerLineFrequency` in `bids_dir/sub-1/ieeg/sub-1_task-mytask_ieeg.json` is null which causes a BIDS validator error.

The dig error looks like it might be related to something @adam2392 was working on recently?
Not sure if anyone knows what happened with powerline and how it could possibly be null instead of 'n/a' but I can look into that later.

### PR DESCRIPTION
**Pull request by alexrockhill**
Fixes #503.

I was going to put this off but if @agramfort can keep up-to-date with every mne issue, I can write a five minute PR right away :)

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

### CODE REVIEW

**Comment by alexrockhill** [comment]
The only thing that's weird is that PowerLineFrequency in the json accompanying an meg/eeg/ieeg file can't be `'n/a'` so I set it to `-1`. Seems like the best option if there is no PowerLineFrequency given but we could also just throw an error that the data must have more than 5 channels and sfreq > 100 or else you'll get an error. But if sfreq < 100, then maybe it makes sense not to provide PowerLineFrequency...

File: mne_bids/utils.py [hunk 1]
```diff
@@ -359,7 +359,7 @@ def _estimate_line_freq(raw, verbose=False):
     if len(picks) < 5:
         warn("Estimation of line frequency only "
              "supports 'meg', 'eeg', 'ecog', or 'seeg'.")
-        return None
+        return -1
```
Comment (agramfort) [inline mne_bids/utils.py:362]: -1 ? are you a C programmer now?

File: mne_bids/tests/test_read.py [hunk 2]
```diff
@@ -266,7 +266,7 @@ def test_line_freq_estimation():
     somato_raw.set_channel_types({somato_raw.ch_names[i]: 'bio'
                                   for i in range(len(somato_raw.ch_names))})
     somato_raw = _handle_info_reading(sidecar_fname, somato_raw, verbose=True)
-    assert somato_raw.info['line_freq'] is None
+    assert somato_raw.info['line_freq'] == -1
```
Comment (agramfort) [inline mne_bids/tests/test_read.py:269]: I would object here. I would keep it None

**Review by agramfort — COMMENTED** [review]
you'll need to fix conflict in what's new too

thx @alexrockhill

File: mne_bids/tests/test_read.py (hunk 2 above)
Comment (alexrockhill) [reply mne_bids/tests/test_read.py:269]: The bids-validator throws an error if it's not an integer, you can't write None or -1 to the file, that's why I did it that way

File: mne_bids/utils.py (hunk 1 above)
Comment (alexrockhill) [reply mne_bids/utils.py]: I moved it to where it was written to the file, I think that's better. Good one ;)

**Comment by alexrockhill** [comment]
Ok should be good to go once tests pass

File: mne_bids/write.py [hunk 3]
```diff
@@ -492,9 +492,13 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, overwrite=False,
     powerlinefrequency = raw.info.get('line_freq', None)
     if powerlinefrequency is None:
         powerlinefrequency = _estimate_line_freq(raw)
-        warn('No line frequency found, defaulting to {} Hz '
-             'estimated from multi-taper FFT '
-             'on 10 seconds of data.'.format(powerlinefrequency))
+        if powerlinefrequency is None:
+            powerlinefrequency = -1  # must be int for validator
+            warn('No line frequency found')
```
Comment (jasmainak) [inline mne_bids/write.py:497]: you don't test this

Comment (alexrockhill) [reply mne_bids/write.py:497]: But CodeCov says I have 100% coverage ;)

Fixing it now

**Comment by agramfort** [comment]
Can you also check that the reader does the right thing with line freq when set to -1? I don’t want a -1 to appear in an info :)

**Comment by alexrockhill** [comment]
> Can you also check that the reader does the right thing with line freq when set to -1? I don’t want a -1 to appear in an info :)

Good catch! In testing this, I noticed that when you write with `ch_type='seeg'` you have to read in with `kind='ieeg'`. A bit confusing I think. Is that by design or should that be fixed? cc @jasmainak

**Comment by jasmainak** [comment]
`kind` is a BIDS thing whereas `ch_type` is an MNE channel type

**Comment by jasmainak** [comment]
your commit signatures don't have the right email or the email is not added to your github profile

**Comment by jasmainak** [comment]
also never tag anyone in a commit, it gets really annoying when you get notifications for every rebase. Just speaking from experience :P

**Comment by alexrockhill** [comment]
> your commit signatures don't have the right email or the email is not added to your github profile

Uh oh, fixing that asap

**Comment by alexrockhill** [comment]
> also never tag anyone in a commit, it gets really annoying when you get notifications for every rebase. Just speaking from experience :P

Oops, hopefully that fixed it

**Comment by alexrockhill** [comment]
> > your commit signatures don't have the right email or the email is not added to your github profile
> 
> Uh oh, fixing that asap

Ok I think I fixed the email thing, not sure I can retroactively change anything though.

**Comment by jasmainak** [comment]
> Ok I think I fixed the email thing, not sure I can retroactively change anything though.

it's still not linking to your profile ...

File: mne_bids/tests/test_write.py [hunk 4]
```diff
@@ -203,6 +203,28 @@ def test_get_anonymization_daysback():
             get_anonymization_daysback([raw, raw2, raw4])
 
 
+def test_create(_bids_validate):
```
Comment (jasmainak) [inline mne_bids/tests/test_write.py:206]: the function name is not very descriptive

**Comment by alexrockhill** [comment]
Did that fix the email thing pushing a new commit?

File: doc/whats_new.rst [hunk 5]
```diff
@@ -54,6 +54,7 @@ Bug
 - :func:`mne_bids.read_raw_bids` now populates the list of bad channels based on ``*_channels.tsv`` if (and only if) a ``status`` column is present, ignoring similar metadata stored in raw file (which will still be used if **no** ``status`` column is present in ``*_channels.tsv``), by `Richard Höchenberger`_ (`#499 <https://github.com/mne-tools/mne-bids/pull/499>`_)
 - Ensure that ``Raw.info['bads']`` returned by :func:`mne_bids.read_raw_bids` is always a list, by `Richard Höchenberger`_ (`#501 <https://github.com/mne-tools/mne-bids/pull/501>`_)
 - :func:`mne_bids.write_raw_bids` now ensures that **all** parts of the :class:`mne.io.Raw` instance stay in sync when using anonymization to shift dates, e.g. ``raw.annotations``, by `Richard Höchenberger`_ (`#504 <https://github.com/mne-tools/mne-bids/pull/504>`_)
+- Fix :func:`mne_bids.write_raw_bids` failed BIDS validator for ``raw.info['dig'] = []`` and for data that has too few channels to find powerline info, by `Alex Rockhill`_ (`#505 <https://github.com/mne-tools/mne-bids/pull/505>`_)
```
Comment (jasmainak) [inline doc/whats_new.rst:57]: too few channels to find powerline info?

or too short recording?

Comment (alexrockhill) [reply doc/whats_new.rst:57]: Too few actually, there's no limit on how short the recording can be

Comment (alexrockhill) [reply doc/whats_new.rst:57]: mne_bids.utils line 366
```
tmax = int(min(len(raw.times), 10 * sfreq))
```

File: mne_bids/read.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ def _handle_info_reading(sidecar_fname, raw, verbose=None):
 
     # read in the sidecar JSON's line frequency
     line_freq = sidecar_json.get("PowerLineFrequency")
-    if line_freq == "n/a":
+    if line_freq in ("n/a", -1):
```
Comment (agramfort) [inline mne_bids/read.py:112]: still I am surprise that the standard says that line_freq should be set to -1 and not 'n/a' when not available. Maybe it's more a validator issue?

@sappelhoff thoughts?

**Comment by agramfort** [comment]
@alexrockhill ok for you?

File: doc/whats_new.rst (hunk 5 above)
Comment (jasmainak) [reply doc/whats_new.rst:57]: why can't you estimate line freq if there are too few channels?

**Comment by jasmainak** [comment]
agree with @agramfort on `'n/a'`

**Comment by agramfort** [comment]
I guess you mean too few samples?

>

**Comment by alexrockhill** [comment]
mne_bids.utils line 347

```
# setup picks of the data to get at least 5 channels
pick_dict = {"meg": True}
picks = list(pick_types(raw.info, exclude='bads', **pick_dict))
if len(picks) < 5:
    pick_dict = {"eeg": True}
    picks = pick_types(raw.info, exclude='bads', **pick_dict)
if len(picks) < 5:
    pick_dict = {"ecog": True}
    picks = pick_types(raw.info, exclude='bads', **pick_dict)
if len(picks) < 5:
    pick_dict = {"seeg": True}
    picks = pick_types(raw.info, exclude='bads', **pick_dict)
if len(picks) < 5:
    warn("Estimation of line frequency only "
         "supports 'meg', 'eeg', 'ecog', or 'seeg'.")
    return None
```

I don't know why someone decided they needed five channels but this is in utils

**Comment by jasmainak** [comment]
@alexrockhill for the future it's easier if you provide links :-)

ping @adam2392 what's going on here?

**Comment by adam2392** [comment]
> mne_bids.utils line 347
> 
> ```
> # setup picks of the data to get at least 5 channels
> pick_dict = {"meg": True}
> picks = list(pick_types(raw.info, exclude='bads', **pick_dict))
> if len(picks) < 5:
>     pick_dict = {"eeg": True}
>     picks = pick_types(raw.info, exclude='bads', **pick_dict)
> if len(picks) < 5:
>     pick_dict = {"ecog": True}
>     picks = pick_types(raw.info, exclude='bads', **pick_dict)
> if len(picks) < 5:
>     pick_dict = {"seeg": True}
>     picks = pick_types(raw.info, exclude='bads', **pick_dict)
> if len(picks) < 5:
>     warn("Estimation of line frequency only "
>          "supports 'meg', 'eeg', 'ecog', or 'seeg'.")
>     return None
> ```
> 
> I don't know why someone decided they needed five channels but this is in utils

I suppose that was arbitrary because the "estimation" of line frequency was using a spectral decomp and then taking the median over multiple channels. I just did some arbitrary splicing of the data and I found picking random 5 channels was okay for estimating the line freq on the test datasets.

I'm okay changing that if needed.

**Comment by alexrockhill** [comment]
How about just `min([len(picks), 5])` to save computation and allow less than 5 as was done with time?

**Comment by adam2392** [comment]
> How about just `min([len(picks), 5])` to save computation and allow less than 5 as was done with time?

That sounds fine. The only thing is, the estimation of line freq might not be that great anymore. Maybe try and see if the unit tests still pass.

I think the estimation procedure could be a bit more robust, so hopefully a future iteration can improve it...

**Comment by alexrockhill** [comment]
Ok so an issue is that the usa psd is just over 0.1 standard deviation greater than the eu psd for the truncated sample data. This really complicates rejecting undeterminable frequencies because for the random raw data, the difference by chance is about the same order of magnitude. Looking into it...

**Comment by alexrockhill** [comment]
Ah, if the tests are asymmetric due to the 1 / f bias, that fixes it.

**Comment by alexrockhill** [comment]
Ok nevermind, they're all within margin error by chance.

**Comment by jasmainak** [comment]
actually why are CIs unhappy? Can you make the CIs pass first?

**Comment by alexrockhill** [comment]
So there is no way to fail to find a powerline frequency for random data as written because the test is just whether 50 Hz psd is greater than 60 Hz but if we ask for even the slightest difference to make a conclusion and otherwise fail to make a conclusion, it will happen by chance in the random data regularly greater than the difference in real data. So I added some line noise to the created data and I think the solution as it was is best, just warn the user and rely on a small difference and assume they're not providing random data.

**Comment by alexrockhill** [comment]
I don't know though, if it's not that robust and someone saves a large dataset and filters based on the line noise you incorrectly gave them and then gets conclusions based on differences in line noise that would be very bad. I think I'd lean toward getting rid of estimate line noise altogether and only use it if it's provided in `raw.info['line_freq']`.

**Comment by agramfort** [comment]
+1 I agree as I see that it can silently fail and mne-bids should be
as little computation as possible.

**Comment by alexrockhill** [comment]
I don't see the big upside on estimating the powerline noise, if it's there and it's listed as 'n/a' in the sidecar, the person analyzing the dataset will say well I should get rid of that noise that I see at 60 Hz. But if the json says the powerline noise is at 50 Hz and it's actually at 60 Hz then the user isn't that out-of-line for assuming that they've taken care of powerline noise by filtering around 50 Hz...

**Comment by agramfort** [comment]
Lets merge merge this as is. Unless the coverage drop is legit please merge

**Comment by adam2392** [comment]
> okay fair enough arguments for me. okay with you @adam2392 ?

Sounds fine to me.

**Comment by jasmainak** [comment]
what happened to commits by @agramfort ?

File: mne_bids/utils.py [hunk 7]
```diff
@@ -365,29 +362,25 @@ def _estimate_line_freq(raw, verbose=False):
...
     if verbose is True:
         print("EU (i.e. 50 Hz) PSD is {} and "
               "USA (i.e. 60 Hz) PSD is {}".format(eu_psd, usa_psd))
 
-    if usa_psd > eu_psd:
-        line_freq = 60
-    else:
-        line_freq = 50
-    return line_freq
+    return 60 if usa_psd > eu_psd else 50
```
Comment (agramfort) [inline mne_bids/utils.py:383]: @alexrockhill I will proceed as is but all the changes is this file were here unnecessary.

**Comment by agramfort** [comment]
thx @alexrockhill 

try to avoid squashing commits in a PR as we loose history of contributions including my commits :)

File: mne_bids/read.py (hunk 6 above)
Comment (sappelhoff) [reply mne_bids/read.py]: that's definitely a validator issue. when missing, it should be `n/a`

very weird that the validator wants `-1`. @alexrockhill could you open an issue for that on the validator?

https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/03-electroencephalography.html#sidecar-json-_eegjson

**Pull request merged**